Add Apptainer installation for use on HPC systems (#4857)
spinalcordtoolbox/spinalcordtoolbox: documentation/source/user_section/installation/linux.rst

.. code:: sh

    apt install gcc

On CentOS/RedHat:

.. code:: sh

    yum -y install gcc

Installation Options
====================

Depending on your use case, there are a number of different installation methods available:

- :ref:`Packaged Installation <native-install>`: For general use.
- :ref:`GitHub Installation <github-install>`: For developers.

> - :ref:`Docker Installation <docker-install>`: For Cross-Platform Use/Testing.

Reviewer: Tiny nit (sentence case)

```suggestion
- :ref:`Docker Installation <docker-install>`: For cross-platform use/testing.
```

Author: Resolved with recent batch of commits.

> - **[Experimental]** :ref:`Apptainer Installation <apptainer-install>`: For HPC use and other restricted permission applications.

Reviewer: Maybe it would be good to also use the word "cluster" here just because it gives the user more keywords to go "aha, this is my use-case!"

```suggestion
- **[Experimental]** :ref:`Apptainer Installation <apptainer-install>`: For HPC clusters and other restricted permission applications.
```

> - **[Experimental]** :ref:`PIP Installation <pip-install>` Installation ``pip`` for use as a Python package.

Reviewer: I love this table of contents!! My only concern is that it's a little buried, since the "Requirements" section pushes it off the page. It would be nice if the TOC was front and center upon loading the page?

So, maybe it could be nice to move the TOC up, but include the first headings as well.

Something like:

```
**********************
Installation for Linux
**********************

Depending on your use case, there are a number of different installation methods available:

- :ref:`Requirements <install-requirements>`: Prerequisites for all installations
- Installation Options
   - :ref:`Packaged Installation <native-install>`: For general use.
   - :ref:`GitHub Installation <github-install>`: For developers.
   - :ref:`Docker Installation <docker-install>`: For Cross-Platform Use/Testing.
   - **[Experimental]** :ref:`Apptainer Installation <apptainer-install>`: For HPC use and other restricted permission applications.
   - **[Experimental]** :ref:`PIP Installation <pip-install>` Installation ``pip`` for use as a Python package.

.. warning::

    If you use Windows Subsystem for Linux (WSL), please refer to the :ref:`Windows installation section <wsl-installation>`.
```

Author: Resolved with [df1a199](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4857/commits/df1a199a85ea57fb274bfd283ed53577524b9e34)

I kept the WSL warning front and center, though, to ensure it gets acknowledged by users before they try any of the other install methods.

.. _native-install:

> Install from Package
> ------------------------------

Reviewer: Tiny tiny nit, but generally Sphinx/rST wants us to keep the heading underlines the same length at the header text. (It's a weird restriction!)

```suggestion
Install from Package
--------------------
```

Author: Ah, I knew I would miss one. Thanks for catching it.

> The simplest way to install SCT, installing an in-place, static version of a tested package release. If you do not have any special circumstances, we recommend using this installation method.

Reviewer: Tiny wording nit:

```suggestion
The simplest way to install SCT is to use an in-place, static version of a tested package release. If you do not have any special circumstances, we recommend using this installation method.
```

Reviewer: (Thanks for adding details here btw! Noting that it's a tested release is A+.)

Then, to forward the X11 server:

.. note::

    The following instructions have been tested with Xorg and xWayland.

    Set up may vary if you are using a different X11 server.

#. Install ``xauth`` and ``xhost`` on the host machine, if not already installed:

    For example on Debian/Ubuntu:

    .. code:: bash

        sudo apt install xauth x11-xserver-utils

#. Permit docker access to the X11 Server

   If hosting container from the local machine:

    .. code:: bash

        xhost +local:docker

#. In your Terminal window, run:
   
    .. code:: bash

        sudo docker run -it --rm --privileged -e DISPLAY=$DISPLAY -v /tmp/.X11-unix:/tmp/.X11-unix <IMAGE_NAME>/ubuntu:ubuntu22.04``

You can test whether GUI scripts are available by running the following command in your Docker container:
 
    .. code:: bash
   
        sct_check_dependencies
      
You should see two green ``[OK]`` symbols at the bottom of the report for "PyQT" and "matplotlib" checks, which represent the GUI features provided by SCT are now available.

.. _apptainer-install:

**[EXPERIMENTAL]** Install within Apptainer
-------------------------------------------

> Like Docker, `Apptainer <https://apptainer.org/docs/user/main/introduction.html>`_ is a portable container platform. It was designed with a focus on being used in "shared system" contexts, where multiple users with different needs require access to the same hardware. If you need to run SCT in this context (and a native install is not possible, as is often the case inHigh Performance Computer (HPC) clusters), you should install SCT in this way.

Reviewer: Might be good to also mention Singularity here, too:

```suggestion
Like Docker, `Apptainer <https://apptainer.org/docs/user/main/introduction.html>`_ (formerly Singularity) is a portable container platform. It was designed with a focus on being used in "shared system" contexts, where multiple users with different needs require access to the same hardware. If you need to run SCT in this context (and a native install is not possible, as is often the case in High Performance Computer (HPC) clusters), you should install SCT in this way.
```

Using Apptainer introduces a few caveats to using SCT, however:

- Apptainer containers will only work on Linux-based systems, and cannot be ran on Windows or MacOS.

> - Due to containers being static post-creation, functions which install or modifying SCT (such as ``deepseg -install``) will not work. See `here <apptainer-task-install>`_ for a workaround.

Reviewer: Nit:

```suggestion
- Due to containers being static post-creation, functions which install or modify SCT (such as ``sct_deepseg -install``) will not work. See `here <apptainer-task-install>`_ for a workaround.
```

Reviewer: The link to the `apptainer-task-install` anchor is broken for me, too? It sends me here:

https://spinalcordtoolbox--4857.org.readthedocs.build/4857/user_section/installation/apptainer-task-install

I think maybe we need to use `:ref:` syntax instead, as was done in the TOC.

```suggestion
- Due to containers being static post-creation, functions which install or modify SCT (such as ``sct_deepseg -install``) will not work. See :ref:`here <apptainer-task-install>` for a workaround.
```

Author: Good catches, just resolved these.

Basic Installation
******************

This installation will install the latest SCT version available on the master branch.

#. Install Apptainer if you have not done so already (or activate the module which contains it, if on an shared resource system)

#. Download the requisite files from GitHub:

    .. code:: sh

        curl "raw.githubusercontent.com/spinalcordtoolbox/spinalcordtoolbox/refs/heads/master/contrib/apptainer/sct.def" -o "sct.def"

>         curl "raw.githubusercontent.com/spinalcordtoolbox/spinalcordtoolbox/refs/heads/master/contrib/apptainer/install_sct_containered.sh" -o "install_sct_containered.sh"

Reviewer: Are these the kinds of files we might consider hosting as assets on each release?

Author: Good point; I agree this should be "pinned" to improve reproducibility. However, we don't have a "release" to place the files in at the moment, so I was just writing the documentation to work with the current code base. 

Lets bring it up at the next SCT meeting to ensure everyone is on the same page though.

Author: As discussed at today's meeting, this is absolutely what we should be doing. Will update our release code to reflect this in the next batch of pushes!

Author: The recent batch of Github CI changes (up to [c73e945](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4857/commits/c73e9459619ee80b1b1acc287a7d280618789e2b)) have resolved this, as well as making future changes which require modifying how releases are made much less scary to test.

> #. Run the installation script. You may also provide provide a list of ``deepseg`` tasks you want installed as well:

Reviewer: It might be good to link to the `sct_deepseg` model gallery here too, so that users can get a sense of which models they have to choose from when pre-installing?

```suggestion
#. Run the installation script. You may also provide provide a list of :ref:`sct_deepseg` tasks you want pre-installed as well:
```

Reviewer: One neat thing is that any SCT scripts at all can be linked to their doc page using `:ref:` :)

Author: Resolved w/ [2c06286](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4857/commits/2c0628683331989205ee6e2394526c21d3e14cf4)

    Basic installation (without any ``deepseg`` tasks)

    .. code:: sh

        ./install_sct_containered.sh

    Installing the ``spinalcord`` and ``tumor_t2`` tasks as well:

    .. code:: sh

        ./install_sct_containered.sh spinalcord tumor_t2

If installation ran to completion, without error, a ``sct.sif`` file should now be present in the directory. This can be used to run any SCT command as if SCT were installed locally; just prepend ``apptainer exec sct.sif`` before it. For example, to run a spinal cord segmentation using DeepSeg:

.. code:: sh

    apptainer exec sct.sif sct_deepseg spinalcord -i example_T2w.nii.gz

.. _apptainer-task-install:

Installing DeepSeg Tasks Post-Install
*************************************

> If you need to install a task after the initial ``sct.sif`` file was created, you can use the following instructions. Note, however, that each time you do this, the ``.sif`` file is rebuilt, which can take quite a while to do; try to determine which ``deepseg`` models you'll need as early as possible to avoid this!

Reviewer: Minor wording nit for flow:
```suggestion
If you need to install a task after the initial ``sct.sif`` file was created, you can use the following instructions. Note, however, that each time you do this, the ``.sif`` file is rebuilt, which can take quite a while to do. To avoid this, try to determine which ``deepseg`` models you'll need as early as possible!
```

Author: Resolved w/ [0977062](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4857/commits/0977062415e991b90f9a4c97ccc05fd940f1671a)

Using Apptainer introduces a few caveats to using SCT, however:

- Apptainer containers will only work on Linux-based systems, and cannot be ran on Windows or MacOS.
- Due to containers being static post-creation, functions which install or modify SCT (such as ``sct_deepseg -install``) will not work. See :ref:`here <apptainer-task-install>` for a workaround.

Basic Installation
******************

This installation will install the latest SCT version available on the master branch.

#. Install Apptainer if you have not done so already (or activate the module which contains it, if on an shared resource system)

#. Download the requisite files from GitHub:

    .. code:: sh

        curl "raw.githubusercontent.com/spinalcordtoolbox/spinalcordtoolbox/refs/heads/master/contrib/apptainer/sct.def" -o "sct.def"
        curl "raw.githubusercontent.com/spinalcordtoolbox/spinalcordtoolbox/refs/heads/master/contrib/apptainer/install_sct_containered.sh" -o "install_sct_containered.sh"

#. Run the installation script. You may also provide provide a list of :ref:`sct_deepseg` tasks you want installed as well:

    Basic installation (without any ``sct_deepseg`` tasks)

    .. code:: sh

>         ./install_sct_containered.sh

Author: Huh, strange. Never had that issue myself. I've added a note regarding this in [187e806](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4857/commits/187e806da35bb56dfb17a8c752284143d3ef89c2)
